File: apps/api/data/mysql/rental_transformer.go
package mysql

import "apps/api/domain"

func ToRentalDB(domainRental domain.Rental) Rental {
	return Rental{
		Id:         domainRental.Id,
		Code:       domainRental.Code,
		Name:       domainRental.Name,
		CheckinAt:  domainRental.CheckinAt,
		CheckoutAt: domainRental.CheckoutAt,
		VariantId:  domainRental.VariantId,
		Variant:    ToVariantDB(domainRental.Variant),
		CreatedAt:  domainRental.CreatedAt,
		DeletedAt:  domainRental.DeletedAt,
	}
}

func ToRentalDomain(dbRental Rental) domain.Rental {
	return domain.Rental{
		Id:         dbRental.Id,
		Code:       dbRental.Code,
		Name:       dbRental.Name,
		VariantId:  dbRental.VariantId,
		Variant:    ToVariantDomain(dbRental.Variant),
		CheckinAt:  dbRental.CheckinAt,
		CheckoutAt: dbRental.CheckoutAt,
		CreatedAt:  dbRental.CreatedAt,
		DeletedAt:  dbRental.DeletedAt,
	}
}

func ToRentalListDomain(dbRentals []Rental) []domain.Rental {
	var domainRentals []domain.Rental
	for _, dbRental := range dbRentals {
		domainRentals = append(domainRentals, ToRentalDomain(dbRental))
	}
	return domainRentals
}

func ToRentalListDB(domainRentals []domain.Rental) []Rental {
	var dbRentals []Rental
	for _, domainRental := range domainRentals {
		dbRentals = append(dbRentals, ToRentalDB(domainRental))
	}
	return dbRentals
}

File: libs/ui/src/data/api/rental.transformer.ts
// eslint-disable-next-line @nx/enforce-module-boundaries
import { Rental as ApiRental } from '../../../../api-contract/src';
import { Rental, RentalCheckinForm, RentalCheckoutForm } from '../../domain';
import { toVariant } from './variant.transformer';

export function toRental(rental: ApiRental): Rental {
  return {
    id: rental.id,
    code: rental.code,
    name: rental.name,
    checkinAt: rental.checkinAt,
    checkoutAt: rental.checkoutAt ?? null,
    createdAt: rental.createdAt,
    variant: toVariant(rental.variant),
  };
}

export function toApiRentalCheckin(form: RentalCheckinForm) {
  return form.rentals.map((rental) => ({
    code: rental.code,
    name: form.name,
    variantId: rental.variant.id,
    checkinAt: form.checkinAt
      ? new Date(
          form.checkinAt.year,
          form.checkinAt.month,
          form.checkinAt.date,
          form.checkinAt.hour,
          form.checkinAt.minute,
          0,
          0
        ).toISOString()
      : new Date().toISOString(),
  }));
}

export function toApiRentalCheckout(form: RentalCheckoutForm) {
  return form.rentals.map((rental) => rental.id);
}
